File: python/flyql/columns/validator.py
"""Columns validator producing positioned diagnostics.

Validates parsed column expressions against a column schema and
transformer registry, returning Diagnostic objects with source ranges.
"""

from typing import Any, List, Optional

from flyql.core.column import Column, ColumnSchema
from flyql.core.range import Range
from flyql.core.validator import (
    CODE_ARG_COUNT,
    CODE_ARG_TYPE,
    CODE_CHAIN_TYPE,
    CODE_RENDERER_ARG_COUNT,
    CODE_RENDERER_ARG_TYPE,
    CODE_UNKNOWN_COLUMN,
    CODE_UNKNOWN_RENDERER,
    CODE_UNKNOWN_TRANSFORMER,
    Diagnostic,
    make_diag,
)
from flyql.flyql_type import Type, type_permits_unknown_children
from flyql.renderers.registry import (
    RendererRegistry,
    default_registry as default_renderer_registry,
)
from flyql.transformers.registry import TransformerRegistry, default_registry

from .column import ParsedColumn


def _python_to_flyql_type(v: Any) -> Optional[Type]:
    # bool check MUST precede int check (bool is subclass of int in Python)
    if isinstance(v, bool):
        return Type.Bool
    if isinstance(v, int):
        return Type.Int
    if isinstance(v, float):
        return Type.Float
    if isinstance(v, str):
        return Type.String
    return None


def diagnose(
    parsed_columns: List[ParsedColumn],
    schema: ColumnSchema,
    registry: Optional[TransformerRegistry] = None,
    renderer_registry: Optional[RendererRegistry] = None,
) -> List[Diagnostic]:
    if not parsed_columns:
        return []
    if registry is None:
        registry = default_registry()
    if renderer_registry is None:
        renderer_registry = default_renderer_registry()

    diags: List[Diagnostic] = []

    for col in parsed_columns:
        segments = col.segments
        if segments and segments[-1] == "":
            segments = segments[:-1]
        if not segments:
            continue
        resolved = schema.resolve(segments)

        prev_output_type: Optional[Type] = None
        if resolved is None:
            if col.name_range is not None:
                fail_seg, fail_range, parent = _find_failing_segment(
                    col, schema, col.name_range, segments
                )
                if parent is not None and type_permits_unknown_children(parent.type):
                    # Permissive parent (JSON, JSONString, Map, Unknown) — undeclared
                    # nested key access is allowed; suppress diag and treat downstream
                    # input type as unknown.
                    prev_output_type = None
                else:
                    diags.append(
                        make_diag(
                            range=fail_range,
                            message=f"column '{fail_seg}' is not defined",
                            severity="error",
                            code=CODE_UNKNOWN_COLUMN,
                        )
                    )
        elif resolved.type != Type.Unknown:
            prev_output_type = resolved.type

        transformer_ranges = col.transformer_ranges or []

        for ti, transformer in enumerate(col.transformers):
            ranges = transformer_ranges[ti] if ti < len(transformer_ranges) else {}
            name_range = ranges.get("name_range")
            arg_ranges = ranges.get("argument_ranges", [])
            t = registry.get(transformer["name"])

            if t is None:
                if name_range is not None:
                    diags.append(
                        make_diag(
                            range=name_range,
                            message=f"unknown transformer: '{transformer['name']}'",
                            severity="error",
                            code=CODE_UNKNOWN_TRANSFORMER,
                        )
                    )
                prev_output_type = None
                continue

            required_count = sum(
                1 for s in t.arg_schema if getattr(s, "required", True)
            )
            max_count = len(t.arg_schema)
            got = len(transformer["arguments"])
            if got < required_count or got > max_count:
                if required_count == max_count:
                    expect_str = f"{required_count} arguments"
                else:
                    expect_str = f"{required_count}..{max_count} arguments"
                if name_range is not None:
                    full_range = name_range
                    if arg_ranges:
                        full_range = Range(
                            name_range.start,
                            arg_ranges[-1].end + 1,
                        )
                    diags.append(
                        make_diag(
                            range=full_range,
                            message=f"{transformer['name']} expects {expect_str}, got {got}",
                            severity="error",
                            code=CODE_ARG_COUNT,
                        )
                    )

            for j, arg in enumerate(transformer["arguments"]):
                if j >= len(t.arg_schema):
                    break
                expected = t.arg_schema[j].type
                actual = _python_to_flyql_type(arg)
                if actual is None:
                    continue
                if actual == expected:
                    continue
                if actual == Type.Int and expected == Type.Float:
                    continue
                if j < len(arg_ranges):
                    diags.append(
                        make_diag(
                            range=arg_ranges[j],
                            message=f"argument {j + 1} of {transformer['name']}: expected {expected}, got {actual}",
                            severity="error",
                            code=CODE_ARG_TYPE,
                        )
                    )

            if (
                prev_output_type is not None
                and t.input_type is not Type.Any
                and prev_output_type != t.input_type
            ):
                if name_range is not None:
                    diags.append(
                        make_diag(
                            range=name_range,
                            message=f"{transformer['name']} expects {t.input_type} input, got {prev_output_type}",
                            severity="error",
                            code=CODE_CHAIN_TYPE,
                        )
                    )

            prev_output_type = t.output_type

        renderer_ranges = col.renderer_ranges or []
        for ri, renderer in enumerate(col.renderers):
            r_ranges = renderer_ranges[ri] if ri < len(renderer_ranges) else {}
            r_name_range = r_ranges.get("name_range")
            r_arg_ranges = r_ranges.get("argument_ranges", [])
            r = renderer_registry.get(renderer["name"])

            if r is None:
                if r_name_range is not None:
                    diags.append(
                        make_diag(
                            range=r_name_range,
                            message=f"unknown renderer: '{renderer['name']}'",
                            severity="error",
                            code=CODE_UNKNOWN_RENDERER,
                        )
                    )
                continue

            required_count = sum(
                1 for s in r.arg_schema if getattr(s, "required", True)
            )
            max_count = len(r.arg_schema)
            got = len(renderer["arguments"])
            if got < required_count or got > max_count:
                if required_count == max_count:
                    expect_str = f"{required_count} arguments"
                else:
                    expect_str = f"{required_count}..{max_count} arguments"
                if r_name_range is not None:
                    full_range = r_name_range
                    if r_arg_ranges:
                        full_range = Range(
                            r_name_range.start,
                            r_arg_ranges[-1].end + 1,
                        )
                    diags.append(
                        make_diag(
                            range=full_range,
                            message=f"{renderer['name']} expects {expect_str}, got {got}",
                            severity="error",
                            code=CODE_RENDERER_ARG_COUNT,
                        )
                    )

            for j, arg in enumerate(renderer["arguments"]):
                if j >= len(r.arg_schema):
                    break
                expected = r.arg_schema[j].type
                actual = _python_to_flyql_type(arg)
                if actual is None:
                    continue
                if actual == expected:
                    continue
                if actual == Type.Int and expected == Type.Float:
                    continue
                if j < len(r_arg_ranges):
                    diags.append(
                        make_diag(
                            range=r_arg_ranges[j],
                            message=f"argument {j + 1} of {renderer['name']}: expected {expected}, got {actual}",
                            severity="error",
                            code=CODE_RENDERER_ARG_TYPE,
                        )
                    )

            hook_diags = r.diagnose(renderer["arguments"], col)
            if hook_diags:
                diags.extend(hook_diags)

        chain_hook = renderer_registry.get_diagnose()
        if chain_hook is not None and col.renderers:
            chain_diags = chain_hook(col, col.renderers)
            if chain_diags:
                diags.extend(chain_diags)

    return diags


def _find_failing_segment(
    col: ParsedColumn,
    schema: ColumnSchema,
    name_range: Range,
    segments: Optional[List[str]] = None,
) -> tuple[str, Range, Optional[Column]]:
    """Find the first unresolvable segment and its source range.

    Also returns the deepest *resolved* parent column (or None when the
    failing segment is the root). Callers inspect ``parent.type`` against
    :func:`type_permits_unknown_children` to decide whether to suppress
    the unknown-column diagnostic for paths under JSON-family parents.
    """
    segs = segments if segments is not None else col.segments
    current: Optional[Column] = None
    previous: Optional[Column] = None
    for i, seg in enumerate(segs):
        # TD-10 strict pattern: capture previous BEFORE any reassignment of
        # current. Both the lookup-hit branch and the no-children else
        # branch reset current — without capturing previous as the FIRST
        # statement of the loop body, the permissive-parent check at the
        # call site would see None for any failure where the parent has
        # no children at all.
        previous = current
        if i == 0:
            current = schema.get(seg)
        elif current is not None and current.children is not None:
            current = current.children.get(seg.lower())
        else:
            current = None
        if current is None:
            offset = name_range.start
            for j in range(i):
                offset += len(segs[j]) + 1  # +1 for dot
            return seg, Range(offset, offset + len(seg)), previous
    return segs[0], Range(name_range.start, name_range.start + len(segs[0])), None

File: javascript/packages/flyql/src/columns/validator.js
import {
    Diagnostic,
    CODE_UNKNOWN_COLUMN,
    CODE_UNKNOWN_TRANSFORMER,
    CODE_UNKNOWN_TRANSFORMER_ARG_COLUMN,
    CODE_ARG_COUNT,
    CODE_ARG_TYPE,
    CODE_CHAIN_TYPE,
    CODE_UNKNOWN_RENDERER,
    CODE_RENDERER_ARG_COUNT,
    CODE_RENDERER_ARG_TYPE,
    jsToFlyQLType,
    makeDiag,
} from '../core/validator.js'
import { Range } from '../core/range.js'
import { Type } from '../flyql_type.js'
import { defaultRegistry } from '../transformers/registry.js'
import { defaultRegistry as defaultRendererRegistry } from '../renderers/registry.js'

export function diagnose(parsedColumns, schema, registry = null, rendererRegistry = null) {
    if (!parsedColumns || parsedColumns.length === 0) return []
    if (registry == null) registry = defaultRegistry()
    if (rendererRegistry == null) rendererRegistry = defaultRendererRegistry()

    const diags = []

    for (const col of parsedColumns) {
        const rawSegments = col.segments || col.name.split('.')
        const segments =
            rawSegments.length > 0 && rawSegments[rawSegments.length - 1] === ''
                ? rawSegments.slice(0, -1)
                : rawSegments
        if (segments.length === 0) continue
        const resolved = schema.resolve(segments)

        let prevOutputType
        if (resolved == null) {
            if (col.nameRange) {
                const { segment, range } = _findFailingSegment(col, schema, segments)
                diags.push(makeDiag(range, `column '${segment}' is not defined`, 'error', CODE_UNKNOWN_COLUMN))
            }
            prevOutputType = null
        } else {
            prevOutputType = resolved.type && resolved.type !== Type.Unknown ? resolved.type : null
        }

        const transformerRanges = col.transformerRanges || []

        for (let ti = 0; ti < col.transformers.length; ti++) {
            const transformer = col.transformers[ti]
            const ranges = transformerRanges[ti] || {}
            const t = registry.get(transformer.name)

            if (t == null) {
                if (ranges.nameRange) {
                    diags.push(
                        makeDiag(
                            ranges.nameRange,
                            `unknown transformer: '${transformer.name}'`,
                            'error',
                            CODE_UNKNOWN_TRANSFORMER,
                        ),
                    )
                }
                prevOutputType = null
                continue
            }

            const requiredCount = t.argSchema.filter((s) => s.required).length
            const maxCount = t.argSchema.length
            const got = transformer.arguments.length
            if (got < requiredCount || got > maxCount) {
                let expectStr
                if (requiredCount === maxCount) {
                    expectStr = `${requiredCount} arguments`
                } else {
                    expectStr = `${requiredCount}..${maxCount} arguments`
                }
                if (ranges.nameRange) {
                    const argRangesArr = ranges.argumentRanges || []
                    const fullRange =
                        argRangesArr.length > 0
                            ? new Range(ranges.nameRange.start, argRangesArr[argRangesArr.length - 1].end + 1)
                            : ranges.nameRange
                    diags.push(
                        makeDiag(
                            fullRange,
                            `${transformer.name} expects ${expectStr}, got ${got}`,
                            'error',
                            CODE_ARG_COUNT,
                        ),
                    )
                }
            }

            const argRanges = ranges.argumentRanges || []
            const transformerArgKinds = transformer.argumentKinds || []
            // Guard against parser-bug or legacy-AST drift: only trust kinds when
            // length matches the argument list. On mismatch fall through to the
            // pre-existing literal type checks.
            const transformerKindsInSync = transformerArgKinds.length === transformer.arguments.length
            for (let j = 0; j < transformer.arguments.length; j++) {
                if (j >= t.argSchema.length) break
                if (transformerKindsInSync && transformerArgKinds[j] === 'col') {
                    const rawRef = transformer.arguments[j]
                    if (
                        typeof rawRef === 'string' &&
                        schema.resolve(rawRef.split('.')) == null &&
                        j < argRanges.length
                    ) {
                        diags.push(
                            makeDiag(
                                argRanges[j],
                                `unknown column in argument: '${rawRef}'`,
                                'error',
                                CODE_UNKNOWN_TRANSFORMER_ARG_COLUMN,
                            ),
                        )
                    }
                    continue
                }
                const expected = t.argSchema[j].type
                const actual = jsToFlyQLType(transformer.arguments[j])
                if (actual == null) continue
                if (actual === expected) continue
                if (actual === Type.Int && expected === Type.Float) continue
                if (j < argRanges.length) {
                    diags.push(
                        makeDiag(
                            argRanges[j],
                            `argument ${j + 1} of ${transformer.name}: expected ${expected}, got ${actual}`,
                            'error',
                            CODE_ARG_TYPE,
                        ),
                    )
                }
            }

            if (prevOutputType != null && prevOutputType !== t.inputType) {
                if (ranges.nameRange) {
                    diags.push(
                        makeDiag(
                            ranges.nameRange,
                            `${transformer.name} expects ${t.inputType} input, got ${prevOutputType}`,
                            'error',
                            CODE_CHAIN_TYPE,
                        ),
                    )
                }
            }

            prevOutputType = t.outputType
        }

        const rendererRanges = col.rendererRanges || []
        for (let ri = 0; ri < col.renderers.length; ri++) {
            const renderer = col.renderers[ri]
            const ranges = rendererRanges[ri] || {}
            const r = rendererRegistry.get(renderer.name)

            if (r == null) {
                if (ranges.nameRange) {
                    diags.push(
                        makeDiag(
                            ranges.nameRange,
                            `unknown renderer: '${renderer.name}'`,
                            'error',
                            CODE_UNKNOWN_RENDERER,
                        ),
                    )
                }
                continue
            }

            const requiredCount = r.argSchema.filter((s) => s.required).length
            const maxCount = r.argSchema.length
            const got = renderer.arguments.length
            if (got < requiredCount || got > maxCount) {
                let expectStr
                if (requiredCount === maxCount) {
                    expectStr = `${requiredCount} arguments`
                } else {
                    expectStr = `${requiredCount}..${maxCount} arguments`
                }
                if (ranges.nameRange) {
                    const argRangesArr = ranges.argumentRanges || []
                    const fullRange =
                        argRangesArr.length > 0
                            ? new Range(ranges.nameRange.start, argRangesArr[argRangesArr.length - 1].end + 1)
                            : ranges.nameRange
                    diags.push(
                        makeDiag(
                            fullRange,
                            `${renderer.name} expects ${expectStr}, got ${got}`,
                            'error',
                            CODE_RENDERER_ARG_COUNT,
                        ),
                    )
                }
            }

            const rArgRanges = ranges.argumentRanges || []
            const rendererArgKinds = renderer.argumentKinds || []
            const rendererKindsInSync = rendererArgKinds.length === renderer.arguments.length
            for (let j = 0; j < renderer.arguments.length; j++) {
                if (j >= r.argSchema.length) break
                if (rendererKindsInSync && rendererArgKinds[j] === 'col') {
                    const rawRef = renderer.arguments[j]
                    if (
                        typeof rawRef === 'string' &&
                        schema.resolve(rawRef.split('.')) == null &&
                        j < rArgRanges.length
                    ) {
                        diags.push(
                            makeDiag(
                                rArgRanges[j],
                                `unknown column in argument: '${rawRef}'`,
                                'error',
                                CODE_UNKNOWN_TRANSFORMER_ARG_COLUMN,
                            ),
                        )
                    }
                    continue
                }
                const expected = r.argSchema[j].type
                const actual = jsToFlyQLType(renderer.arguments[j])
                if (actual == null) continue
                if (actual === expected) continue
                if (actual === Type.Int && expected === Type.Float) continue
                if (j < rArgRanges.length) {
                    diags.push(
                        makeDiag(
                            rArgRanges[j],
                            `argument ${j + 1} of ${renderer.name}: expected ${expected}, got ${actual}`,
                            'error',
                            CODE_RENDERER_ARG_TYPE,
                        ),
                    )
                }
            }

            const hookDiags = r.diagnose ? r.diagnose(renderer.arguments, col, ranges) : []
            if (hookDiags && hookDiags.length > 0) {
                diags.push(...hookDiags)
            }
        }

        const chainHook = rendererRegistry.getDiagnose ? rendererRegistry.getDiagnose() : null
        if (chainHook && col.renderers.length > 0) {
            const chainDiags = chainHook(col, col.renderers)
            if (chainDiags && chainDiags.length > 0) {
                diags.push(...chainDiags)
            }
        }
    }

    return diags
}

function _findFailingSegment(col, schema, segments) {
    let current = null
    for (let i = 0; i < segments.length; i++) {
        if (i === 0) {
            current = schema.get(segments[i])
        } else if (current != null && current.children != null) {
            current = current.children[segments[i].toLowerCase()] || null
        } else {
            current = null
        }
        if (current == null) {
            let offset = col.nameRange.start
            for (let j = 0; j < i; j++) {
                offset += segments[j].length + 1
            }
            return { segment: segments[i], range: new Range(offset, offset + segments[i].length) }
        }
    }
    return { segment: segments[0], range: new Range(col.nameRange.start, col.nameRange.start + segments[0].length) }
}
